app/hooks/search.tsx

```diff
@@ -0,0 +1,171 @@
+
+export type Question = {
+  pageid: string
+  title: string
+}
+export type SearchResult = Question & {
+  score: number
+  model: string
+  url?: string
+}
+
+/**
+ * Sort function for the highest score on top
+ */
+const byScore = (a: SearchResult, b: SearchResult) => b.score - a.score
+
+/** Baseline full-text search matching the query with each question as strings, weighting down:
+ *  short words,
+ *  wh* questions,
+ *  distance,
+ *  partial (prefix) match without full match
+ *  normalized to ignore a/an/the, punctuation, and case
+ */
+export const makeBaselineSearcher =
+  (questions: MutableRefObject<Question[]>, numResults = 5) =>
+  async (searchQueryRaw: string): Promise<SearchResult[]> => {
+    if (!searchQueryRaw) {
+      return []
+    }
+
+    const searchQueryTokens = normalize(searchQueryRaw).split(' ')
+    const matchers = searchQueryTokens.map((token) => ({
+      weight: token.match(/^(?:\w|\w\w|wh.*|how)$/) ? 0.2 : token.length,
+      fullRe: new RegExp(`\\b${token}\\b`),
+      prefixRe: new RegExp(`\\b${token}`),
+    }))
+    const totalWeight = matchers.reduce((acc, {weight}) => acc + weight, 0.1) // extra total to only approach 100%
+
+    const scoringFn = (questionNormalized: string) => {
+      let score = 0
+      let prevPosition = -1
+      for (const {weight, fullRe, prefixRe} of matchers) {
+        const fullMatch = fullRe.exec(questionNormalized)
+        const prefixMatch = prefixRe.exec(questionNormalized)
+        const currPosition = fullMatch?.index ?? prefixMatch?.index ?? prevPosition
+        const distanceMultiplier =
+          questionNormalized.slice(prevPosition, currPosition).split(' ').length === 2 ? 1 : 0.9
+
+        if (fullMatch) {
+          score += weight * distanceMultiplier
+        } else {
+          if (prefixMatch) {
+            score += 0.9 * weight * distanceMultiplier
+          } else {
+            score -= 0.2 * weight
+          }
+        }
+        prevPosition = currPosition
+      }
+
+      return score / totalWeight
+    }
+
+    return questions.current
+      .map(({pageid, title}) => {
+        const normalized = normalize(title)
+        return {
+          pageid,
+          title,
+          normalized,
+          model: 'plaintext',
+          score: scoringFn(normalized),
+        }
+      })
+      .sort(byScore)
+      .slice(0, numResults)
+      .filter(({score}) => score > 0)
+  }
+
+/**
+ * Ignore unimportant details for similarity comparison
+ */
+const normalize = (question: string) =>
+  question
+    .toLowerCase()
+    .replace(/[^\w ]|\b(?:an?|the?)\b/g, '')
+    .replace(/(\w{2})s\b/g, '$1') // cannot use lookbehind (?<=...) because not supported on Safari
+    .replace(/\s+|_|&\s*/g, ' ')
+    .trim()
+
+/**
+ * Construct a semantic search function, that will search over the provided questions
+ */
+const semanticSearcher = async (questions: Question[], encodings: Tensor2D, numResults: number) => {
+  const model = await load()
```

Comment:
on the main thread? can I see a deployed version, please? how is the performance on non-Chrome browsers, especially on mobile phones?

Reply:
https://stampy-ui.ai688.workers.dev/ - firefox works fine, but mobile might be an issue

Reply:
I don't see any search results whatsoever in Firefox on apple-silicon Mac:

<img width="1049" alt="Screenshot 2023-06-08 at 10 47 00" src="https://github.com/StampyAI/stampy-ui/assets/1087670/73dc5303-1974-4c22-a233-0d1666b09e18">

Reply:
they appeared when I checked the tab later, but it's not acceptable to not show anything for ~30 seconds or so